File: mvvmcore/src/main/java/com/handpay/mvvm/BaseVMFragment.java
package com.handpay.mvvm;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

public abstract class BaseVMFragment<VM extends BaseVM> extends Fragment {

    protected VM vm;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(getContentLayoutResId(), container, false);
        initView(rootView);
        return rootView;
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (providerVM() != null)
            vm = ViewModelProviders.of(this).get(providerVM());
        initData();
    }


    protected abstract void initData();

    protected abstract void initView(View view);

    /**
     * ViewModel
     *
     * @return ViewModel
     */
    protected abstract Class<VM> providerVM();

    /**
     * @return layoutId
     */
    @LayoutRes
    protected abstract int getContentLayoutResId();
}

File: mvvmcore/src/main/java/com/handpay/mvvm/BaseVMFragment.kt
package com.handpay.mvvm

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders

abstract class BaseVMFragment<VM : BaseVM?> : Fragment() {
    protected var model: VM? = null
    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(contentLayoutResId, container, false)
        providerVM()?.let {
            model= ViewModelProviders.of(this).get(it)
        }
        initView(rootView)
        return rootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initData()
    }

    protected abstract fun initData()
    protected abstract fun initView(view: View)
    /**
     * ViewModel
     *
     * @return ViewModel
     */
    open fun providerVM(): Class<VM>? = null

    /**
     * @return layoutId
     */
    @get:LayoutRes
    protected abstract val contentLayoutResId: Int
}